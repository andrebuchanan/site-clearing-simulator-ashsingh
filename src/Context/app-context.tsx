import React, { ReactNode, useMemo, useState } from 'react'

type Session = {
  init: boolean
  quit: boolean
}

interface AppContextInterface {
  fuelConsumed: number[]
  itemizedCosts: number[]
  loggedCommands: string[]
  sessionState: Session
  siteData: string[][]
  logCommand: (command: string) => void
  saveSiteData: (data: string[][]) => void
  updateCost: (price: number) => void
  updateFuel: (quantity: number) => void
  updateSession: (session: Session) => void
}

export const AppContext = React.createContext<AppContextInterface>({
  fuelConsumed: [],
  itemizedCosts: [],
  loggedCommands: [],
  sessionState: { init: true, quit: false },
  siteData: [],
  logCommand: () => undefined,
  saveSiteData: () => undefined,
  updateCost: () => undefined,
  updateFuel: () => undefined,
  updateSession: () => undefined
})

export const AppContextProvider = ({ children }: { children: ReactNode }) => {
  const [commands, setCommands] = useState<string[]>([])
  const [cost, setCost] = useState<number[]>([])
  const [fuel, setFuel] = useState<number[]>([])
  const [session, setSession] = useState<Session>({ init: true, quit: false })
  const [siteInput, setSiteInput] = useState<string[][]>([])

  const logCommandHandler = (command: string) => {
    setCommands((prevState) => [...prevState, command])
  }

  const saveSiteDataHandler = (data: string[][]) => {
    setSiteInput(data)
  }

  const updateCostHandler = (price: number) => {
    setCost((prevState) => [...prevState, price])
  }

  const updateFuelHandler = (quantity: number) => {
    setFuel((prevState) => [...prevState, quantity])
  }

  const updateSessionHandler = (sessionProp: Session) => {
    console.log('session', sessionProp)
    if (sessionProp.quit) console.log('quit')
    setSession(sessionProp)
  }

  const contextValue = useMemo(
    () => ({
      fuelConsumed: fuel,
      itemizedCosts: cost,
      loggedCommands: commands,
      sessionState: session,
      siteData: siteInput,
      logCommand: logCommandHandler,
      saveSiteData: saveSiteDataHandler,
      updateCost: updateCostHandler,
      updateFuel: updateFuelHandler,
      updateSession: updateSessionHandler
    }),
    [commands, cost, fuel, session, siteInput]
  )

  return (
    <AppContext.Provider value={contextValue}>{children}</AppContext.Provider>
  )
}

export default AppContextProvider
