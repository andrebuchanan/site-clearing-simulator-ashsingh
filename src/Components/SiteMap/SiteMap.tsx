import React, { useContext, useEffect, useState } from 'react'
import uniqid from 'uniqid'

import { styled } from '@mui/material/styles'
import ArrowBackIcon from '@mui/icons-material/ArrowBack'
import ArrowDownwardIcon from '@mui/icons-material/ArrowDownward'
import ArrowForwardIcon from '@mui/icons-material/ArrowForward'
import ArrowUpwardIcon from '@mui/icons-material/ArrowUpward'
import Button from '@mui/material/Button'
import Grid from '@mui/material/Grid'
import Paper from '@mui/material/Paper'
import Box from '@mui/material/Box'

import { AppContext } from '../../Context'

const Item = styled(Paper)(({ theme }) => ({
  backgroundColor: '#fff',
  ...theme.typography.body2,
  padding: theme.spacing(1),
  textAlign: 'center',
  color: theme.palette.text.secondary
}))

enum Direction {
  Down = 'Down',
  Left = 'Left',
  Right = 'Right',
  Up = 'Up'
}

const bulldozer = (direction: string) => {
  switch (direction) {
    case Direction.Down:
      return <ArrowDownwardIcon />
    case Direction.Left:
      return <ArrowBackIcon />
    case Direction.Right:
      return <ArrowForwardIcon />
    case Direction.Up:
      return <ArrowUpwardIcon />
    default:
      return null
  }
}

export type SiteMapProps = {
  data: string[][]
}

export const SiteMap = ({ data }: SiteMapProps) => {
  const { logCommand } = useContext(AppContext)

  const [begin, setBegin] = useState(true)
  const [active, setActive] = useState({ row: -1, column: 0 })
  const [direction, setDirection] = useState(Direction.Right)

  useEffect(() => {
    const { row, column } = active
    if (!begin && (data[row] || [])[column] === undefined) {
      alert('Quit')
      setActive({ row: -1, column: 0 })
      setBegin(true)
    }
  }, [active])

  const grid = []

  const enterSite = () => {
    setActive((prevState) => ({
      ...prevState,
      row: 0
    }))
    logCommand('Advance: Bulldozer goes Right.')
    setDirection(Direction.Right)
    setBegin(false)
  }

  const goAdvance = () => {
    if (begin) {
      return enterSite()
    }

    switch (direction) {
      case Direction.Down:
        setActive((prevState) => ({
          ...prevState,
          row: prevState.row + 1
        }))
        break
      case Direction.Left:
        setActive((prevState) => ({
          ...prevState,
          column: prevState.column - 1
        }))
        break
      case Direction.Right:
        setActive((prevState) => ({
          ...prevState,
          column: prevState.column + 1
        }))
        break
      case Direction.Up:
        setActive((prevState) => ({
          ...prevState,
          row: prevState.row - 1
        }))
        break
      default:
    }

    logCommand(`Advance: Bulldozer goes ${direction}`)

    return undefined
  }

  const goLeft = () => {
    let message = 'Left: Bulldozer goes '
    switch (direction) {
      case Direction.Down:
        setActive((prevState) => ({
          ...prevState,
          column: prevState.column + 1
        }))
        setDirection(Direction.Right)
        message += Direction.Right
        break
      case Direction.Left:
        setActive((prevState) => ({
          ...prevState,
          row: prevState.row + 1
        }))
        setDirection(Direction.Down)
        message += Direction.Down
        break
      case Direction.Right:
        setActive((prevState) => ({
          ...prevState,
          row: prevState.row - 1
        }))
        setDirection(Direction.Up)
        message += Direction.Up
        break
      case Direction.Up:
        setActive((prevState) => ({
          ...prevState,
          column: prevState.column - 1
        }))
        setDirection(Direction.Left)
        message += Direction.Left
        break
      default:
    }
    logCommand(message)
  }

  const goRight = () => {
    let message = 'Right: Bulldozer goes '
    switch (direction) {
      case Direction.Down:
        setActive((prevState) => ({
          ...prevState,
          column: prevState.column - 1
        }))
        setDirection(Direction.Left)
        message += Direction.Left
        break
      case Direction.Left:
        setActive((prevState) => ({
          ...prevState,
          row: prevState.row - 1
        }))
        setDirection(Direction.Up)
        message += Direction.Up
        break
      case Direction.Right:
        setActive((prevState) => ({
          ...prevState,
          row: prevState.row + 1
        }))
        setDirection(Direction.Down)
        message += Direction.Down
        break
      case Direction.Up:
        setActive((prevState) => ({
          ...prevState,
          column: prevState.column + 1
        }))
        setDirection(Direction.Right)
        message += Direction.Right
        break
      default:
    }
    logCommand(message)
  }

  for (let i = 0; i < data.length; i += 1) {
    for (let j = 0; j < data[0].length; j += 1) {
      grid.push(
        <Grid item xs={1} key={uniqid()}>
          <Item
            sx={{
              backgroundColor:
                active.row === i && active.column === j ? 'green' : '#fff'
            }}
          >
            <>
              {data[i][j]}
              {active.row === i && active.column === j
                ? bulldozer(direction)
                : bulldozer('default')}
            </>
          </Item>
        </Grid>
      )
    }
  }

  return (
    <Box sx={{ width: '100%' }}>
      <Grid
        container
        rowSpacing={1}
        columnSpacing={{ xs: 1, sm: 2, md: 3 }}
        columns={data[0].length}
      >
        {grid}
      </Grid>
      <Button
        data-testid="go-left"
        disabled={begin}
        variant="contained"
        onClick={goLeft}
      >
        Left
      </Button>
      <Button data-testid="go-advance" variant="contained" onClick={goAdvance}>
        Advance
      </Button>
      <Button
        data-testid="go-right"
        disabled={begin}
        variant="contained"
        onClick={goRight}
      >
        Right
      </Button>
    </Box>
  )
}

export default SiteMap
