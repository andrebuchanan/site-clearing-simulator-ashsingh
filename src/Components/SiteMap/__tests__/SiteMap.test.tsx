/* eslint-disable react/jsx-no-constructed-context-values */
import React from 'react'
import { cleanup, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { AppContext } from '../../../Context'

import SiteMap from '../SiteMap'
import mocks from '../__mocks__/test.mock'

const contextValue = {
  fuelConsumed: [],
  itemizedCosts: [],
  loggedCommands: [],
  sessionState: { init: true, quit: false },
  siteData: mocks,
  logCommand: jest.fn,
  saveSiteData: jest.fn,
  updateCost: jest.fn,
  updateFuel: jest.fn,
  updateSession: jest.fn
}

// let componentRender: RenderResult

// beforeEach(() => {
//   console.log('called')
//   componentRender = render(
//     <AppContext.Provider value={contextValue}>
//       <SiteMap />
//     </AppContext.Provider>
//   )
//   return componentRender
// })
afterEach(() => cleanup)

describe('SiteMap', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <AppContext.Provider value={contextValue}>
        <SiteMap />
      </AppContext.Provider>
    )

    expect(screen.getByTestId('go-left').closest('button')).toBeDisabled()
    expect(asFragment()).toMatchSnapshot()
  })

  it('bulldozer enters site', async () => {
    const updateSession = () => {
      contextValue.sessionState.init = false
    }
    render(
      <AppContext.Provider value={{ ...contextValue, updateSession }}>
        <SiteMap />
      </AppContext.Provider>
    )
    await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
    expect(screen.getByTestId('ArrowForwardIcon')).toBeDefined()
    expect(screen.getByTestId('go-left').closest('button')).not.toBeDisabled()
  })

  // it('bulldozer goes advance', async () => {
  //   console.log('testing')
  //   const updateSession = () => {
  //     contextValue.sessionState.init = false
  //   }
  //   render(
  //     <AppContext.Provider value={{ ...contextValue, updateSession }}>
  //       <SiteMap />
  //     </AppContext.Provider>
  //   )
  //   // enter
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   // go down
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-right')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   // go right
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-left')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   // go up
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-left')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   // go left
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-left')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))

  //   expect(screen.getByTestId('ArrowBackIcon')).toBeDefined()
  //   expect(
  //     screen.getByTestId('ArrowBackIcon').closest('div')?.textContent
  //   ).toEqual('o')
  // })

  // it('bulldozer goes right', async () => {
  //   render(<SiteMap data={mocks} />)
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))

  //   await waitFor(() => userEvent.click(screen.getByTestId('go-right')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-right')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-right')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-right')))

  //   expect(screen.getByTestId('ArrowForwardIcon')).toBeDefined()
  //   expect(
  //     screen.getByTestId('ArrowForwardIcon').closest('div')?.textContent
  //   ).toEqual('o')
  // })

  // it('bulldozer goes left', async () => {
  //   render(<SiteMap data={mocks} />)
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-right')))

  //   await waitFor(() => userEvent.click(screen.getByTestId('go-left')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-left')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-left')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-left')))

  //   expect(screen.getByTestId('ArrowDownwardIcon')).toBeDefined()
  //   expect(
  //     screen.getByTestId('ArrowDownwardIcon').closest('div')?.textContent
  //   ).toEqual('o')
  // })

  // it('exits sitemap gracefully, column', async () => {
  //   render(<SiteMap data={mocks} />)
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))

  //   expect(screen.getByTestId('go-left').closest('button')).toBeDisabled()
  // })

  // it('exits sitemap gracefully, row', async () => {
  //   render(<SiteMap data={mocks} />)
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-advance')))
  //   await waitFor(() => userEvent.click(screen.getByTestId('go-left')))

  //   expect(screen.getByTestId('go-left').closest('button')).toBeDisabled()
  // })
})
