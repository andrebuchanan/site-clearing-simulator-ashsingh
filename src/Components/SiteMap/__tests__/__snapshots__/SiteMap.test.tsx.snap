// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteMap renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-8atqhb"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-10mzr36-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          o
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          o
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          t
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          o
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          o
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          t
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          r
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          r
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-gj1fbr-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-a2oq5s-MuiPaper-root"
        >
          r
        </div>
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="go-left"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Left
    </button>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="go-advance"
      tabindex="0"
      type="button"
    >
      Advance
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="go-right"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Right
    </button>
  </div>
</DocumentFragment>
`;
